File: core/controller.py
from __future__ import annotations

import queue
import threading
import time
from typing import Any, Callable, Optional

from .models import AncMode, CodecMode, ConnectionState, DeviceState, EqMode
from .protocol import (
    PacketStreamParser,
    parse_state_update,
    paced_startup_queries,
    query_anc,
    query_eq,
    query_game_mode,
    query_low_latency,
    set_anc,
    set_codec,
    set_eq,
    set_game_mode,
    set_legacy_codec_lhdc,
    set_low_latency,
)
from ..transport.base import BluetoothTransport
from ..transport.windows_rfcomm import WindowsRfcommTransport
from ..transport.windows_serial_spp import WindowsSerialSppTransport
from ..util.device_match import (
    BluetoothDeviceCandidate,
    CandidatePort,
    choose_best_port,
    describe_available_ports,
    describe_available_rfcomm_devices,
    discover_rfcomm_devices,
)
from ..util.logging import configure_logging


class NiceHckDesktopController:
    def __init__(self, event_callback: Callable[[str, dict[str, Any]], None]) -> None:
        self._event_callback = event_callback
        self._state = DeviceState()
        self._parser = PacketStreamParser()
        self._transport: Optional[BluetoothTransport] = None
        self._worker_lock = threading.Lock()
        self._write_lock = threading.Lock()
        self._logger = configure_logging()

    @property
    def state(self) -> DeviceState:
        return self._state

    def auto_connect(self) -> None:
        self._run_async(self._auto_connect)

    def refresh_state(self) -> None:
        self._run_async(self._refresh_state)

    def disconnect(self) -> None:
        self._run_async(self._disconnect)

    def set_anc_mode(self, mode: AncMode) -> None:
        self._run_async(self._set_anc_mode, mode)

    def set_eq_mode(self, mode: EqMode) -> None:
        self._run_async(self._set_eq_mode, mode)

    def set_codec_mode(self, mode: CodecMode) -> None:
        self._run_async(self._set_codec_mode, mode)

    def set_game_mode_enabled(self, enabled: bool) -> None:
        self._run_async(self._set_game_mode_enabled, enabled)

    def set_low_latency_enabled(self, enabled: bool) -> None:
        self._run_async(self._set_low_latency_enabled, enabled)

    def _run_async(self, func: Callable[..., None], *args: Any) -> None:
        threading.Thread(target=lambda: func(*args), daemon=True).start()

    def _auto_connect(self) -> None:
        with self._worker_lock:
            try:
                self._update_state(connection_state=ConnectionState.CONNECTING, last_error=None)

                rfcomm_candidates = discover_rfcomm_devices()
                if rfcomm_candidates:
                    rfcomm_candidate = rfcomm_candidates[0]
                    self._transport = WindowsRfcommTransport(rfcomm_candidate)
                    self._transport.connect()
                    self._transport.start_reader(self._handle_data, self._handle_error)
                    service_name = rfcomm_candidate.service_name or "?"
                    self._update_state(
                        connection_state=ConnectionState.CONNECTED,
                        device_name=rfcomm_candidate.device_name,
                        port_name=f"RFCOMM:{rfcomm_candidate.address}:{service_name}",
                    )
                    self._log(
                        f"已通过 RFCOMM 连接 {rfcomm_candidate.device_name} @ "
                        f"{rfcomm_candidate.address}:{service_name} ({rfcomm_candidate.description})"
                    )
                    self._send_startup_queries()
                    return

                candidate = choose_best_port()
                if candidate is None:
                    rfcomm_lines = describe_available_rfcomm_devices()
                    port_lines = describe_available_ports()
                    detail_parts = []
                    if rfcomm_lines:
                        detail_parts.append("RFCOMM=" + "；".join(rfcomm_lines))
                    if port_lines:
                        detail_parts.append("PORT=" + "；".join(port_lines))
                    detail = " | ".join(detail_parts) if detail_parts else "未检测到任何 RFCOMM 服务或串口"
                    raise RuntimeError(f"未找到可用蓝牙设备。当前检测结果：{detail}")

                self._transport = WindowsSerialSppTransport(candidate)
                self._transport.connect()
                self._transport.start_reader(self._handle_data, self._handle_error)

                self._update_state(
                    connection_state=ConnectionState.CONNECTED,
                    device_name=candidate.device_name,
                    port_name=candidate.port_name,
                )
                self._log(f"已通过串口连接 {candidate.device_name} @ {candidate.port_name} ({candidate.description})")
                self._send_startup_queries()
            except Exception as exc:
                self._handle_error(exc)

    def _disconnect(self) -> None:
        with self._worker_lock:
            if self._transport is not None:
                self._transport.disconnect()
                self._transport = None
            self._parser = PacketStreamParser()
            self._update_state(
                connection_state=ConnectionState.DISCONNECTED,
                device_name="未连接",
                port_name=None,
                left_battery=None,
                right_battery=None,
                case_battery=None,
                selected_codec=None,
            )
            self._log("已断开连接")

    def _refresh_state(self) -> None:
        with self._worker_lock:
            self._ensure_connected()
            self._send_startup_queries()
            self._log("已发送状态查询")

    def _set_anc_mode(self, mode: AncMode) -> None:
        with self._worker_lock:
            self._ensure_connected()
            self._send(set_anc(mode))
            time.sleep(0.1)
            self._send(query_anc())
            self._log(f"已发送 ANC 切换: {mode.label}")

    def _set_eq_mode(self, mode: EqMode) -> None:
        with self._worker_lock:
            self._ensure_connected()
            if not self._state.firmware.supports_extended_eq and mode in {EqMode.FINE, EqMode.VOCAL}:
                raise RuntimeError("当前固件版本过低，不支持该 EQ 模式")
            self._send(set_eq(mode))
            time.sleep(0.1)
            self._send(query_eq())
            self._log(f"已发送 EQ 切换: {mode.label}")

    def _set_codec_mode(self, mode: CodecMode) -> None:
        with self._worker_lock:
            self._ensure_connected()
            if self._state.firmware.supports_modern_codec_switch:
                self._send(set_codec(mode))
            else:
                if mode == CodecMode.SBC:
                    raise RuntimeError("当前固件版本过低，不支持 SBC 编码切换")
                self._send(set_legacy_codec_lhdc(mode == CodecMode.LHDC))
            self._update_state(selected_codec=mode)
            self._log(f"已发送编码切换: {mode.label}")

    def _set_game_mode_enabled(self, enabled: bool) -> None:
        with self._worker_lock:
            self._ensure_connected()
            self._send(set_game_mode(enabled))
            time.sleep(0.1)
            self._send(query_game_mode())
            self._log(f"已发送游戏模式切换: {'开' if enabled else '关'}")

    def _set_low_latency_enabled(self, enabled: bool) -> None:
        with self._worker_lock:
            self._ensure_connected()
            self._send(set_low_latency(enabled))
            time.sleep(0.1)
            self._send(query_low_latency())
            self._log(f"已发送低延迟切换: {'开' if enabled else '关'}")

    def _send_startup_queries(self) -> None:
        for packet in paced_startup_queries():
            self._send(packet)
            time.sleep(0.1)

    def _ensure_connected(self) -> None:
        if self._transport is None or not self._transport.is_connected():
            raise RuntimeError("当前未连接耳机")

    def _send(self, packet: bytes) -> None:
        if self._transport is None:
            raise RuntimeError("蓝牙传输未初始化")
        with self._write_lock:
            self._transport.send(packet)
            self._log(f"发送: {packet.hex(' ').upper()}")

    def _handle_data(self, data: bytes) -> None:
        self._log(f"接收: {data.hex(' ').upper()}")
        for message in self._parser.feed(data):
            update = parse_state_update(message)
            updates: dict[str, Any] = {}
            if update.firmware is not None:
                updates["firmware"] = update.firmware
            if update.left_battery is not None:
                updates["left_battery"] = update.left_battery
            if update.right_battery is not None:
                updates["right_battery"] = update.right_battery
            if update.case_battery is not None:
                updates["case_battery"] = update.case_battery
            if update.anc_mode is not None:
                updates["anc_mode"] = update.anc_mode
            if update.eq_mode is not None:
                updates["eq_mode"] = update.eq_mode
            if update.game_mode_enabled is not None:
                updates["game_mode_enabled"] = update.game_mode_enabled
            if update.low_latency_enabled is not None:
                updates["low_latency_enabled"] = update.low_latency_enabled
            if updates:
                self._log("状态更新: " + ", ".join(f"{key}={value}" for key, value in updates.items()))
                self._update_state(**updates)

    def _handle_error(self, exc: Exception) -> None:
        self._logger.exception("桌面控制器错误", exc_info=exc)
        self._update_state(connection_state=ConnectionState.ERROR, last_error=str(exc))
        self._emit("error", {"message": str(exc)})

    def _update_state(self, **changes: Any) -> None:
        for key, value in changes.items():
            setattr(self._state, key, value)
        self._emit("state", {"state": self._state})

    def _log(self, message: str) -> None:
        self._logger.info(message)
        self._emit("log", {"message": message})

    def _emit(self, event_type: str, payload: dict[str, Any]) -> None:
        self._event_callback(event_type, payload)

File: src/controller.rs
use std::time::Duration;

use thiserror::Error;
use tokio::sync::{mpsc, oneshot};
use tokio::time::{sleep, timeout};
use tracing::{error, info};

use crate::models::{AncMode, CodecMode, ConnectionState, DeviceState, EqMode};
use crate::protocol::{
    PacketStreamParser, ParsedStateUpdate, paced_startup_queries, parse_state_update, query_anc,
    query_eq, query_game_mode, query_low_latency, set_anc, set_codec, set_eq, set_game_mode,
    set_legacy_codec_lhdc, set_low_latency,
};
use crate::transport::{self, Candidate, Connection, ConnectionEvent};

const COMMAND_PACING: Duration = Duration::from_millis(100);
const DISCOVERY_TIMEOUT: Duration = Duration::from_secs(15);
const CONNECTION_TIMEOUT: Duration = Duration::from_secs(6);

#[derive(Clone, Debug)]
pub enum ControllerCommand {
    Connect,
    Disconnect,
    Refresh,
    SetAnc(AncMode),
    SetEq(EqMode),
    SetCodec(CodecMode),
    SetGameMode(bool),
    SetLowLatency(bool),
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub enum ControllerEvent {
    State(DeviceState),
    Log(String),
    Error(String),
}

#[derive(Debug, Error)]
pub enum ControllerError {
    #[error("当前未连接耳机")]
    NotConnected,
    #[error("当前固件版本过低，不支持该 EQ 模式")]
    UnsupportedEq,
    #[error("当前固件版本过低，不支持 SBC 编码切换")]
    UnsupportedCodec,
    #[error("设备发现超时")]
    DiscoveryTimeout,
    #[error("未找到匹配的 NiceHCK 蓝牙设备")]
    NoCandidates,
    #[error("所有连接方式均失败：{0}")]
    AllConnectionsFailed(String),
    #[error(transparent)]
    Transport(#[from] transport::TransportError),
}

#[derive(Clone)]
pub struct ControllerHandle {
    commands: mpsc::UnboundedSender<ActorMessage>,
}

impl ControllerHandle {
    pub fn send(&self, command: ControllerCommand) -> Result<(), ControllerError> {
        self.commands
            .send(ActorMessage::Command(command))
            .map_err(|_| transport::TransportError::Disconnected.into())
    }

    pub async fn shutdown(&self) {
        let (done_tx, done_rx) = oneshot::channel();
        if self.commands.send(ActorMessage::Shutdown(done_tx)).is_ok() {
            let _ = done_rx.await;
        }
    }
}

enum ActorMessage {
    Command(ControllerCommand),
    Shutdown(oneshot::Sender<()>),
}

pub fn spawn_controller() -> (ControllerHandle, mpsc::UnboundedReceiver<ControllerEvent>) {
    let (command_tx, command_rx) = mpsc::unbounded_channel();
    let (event_tx, event_rx) = mpsc::unbounded_channel();
    tokio::spawn(run_actor(command_rx, event_tx));
    (
        ControllerHandle {
            commands: command_tx,
        },
        event_rx,
    )
}

struct ControllerActor {
    state: DeviceState,
    parser: PacketStreamParser,
    connection: Option<Connection>,
    events: mpsc::UnboundedSender<ControllerEvent>,
}

async fn run_actor(
    mut commands: mpsc::UnboundedReceiver<ActorMessage>,
    events: mpsc::UnboundedSender<ControllerEvent>,
) {
    let mut actor = ControllerActor {
        state: DeviceState::default(),
        parser: PacketStreamParser::default(),
        connection: None,
        events,
    };
    actor.emit_state();

    loop {
        tokio::select! {
            message = commands.recv() => match message {
                Some(ActorMessage::Command(command)) => actor.handle_command(command).await,
                Some(ActorMessage::Shutdown(done)) => {
                    actor.disconnect(false).await;
                    let _ = done.send(());
                    break;
                }
                None => {
                    actor.disconnect(false).await;
                    break;
                }
            },
            event = receive_transport_event(&mut actor.connection), if actor.connection.is_some() => {
                actor.handle_transport_event(event).await;
            }
        }
    }
}

async fn receive_transport_event(connection: &mut Option<Connection>) -> Option<ConnectionEvent> {
    match connection {
        Some(connection) => connection.recv().await,
        None => std::future::pending().await,
    }
}

impl ControllerActor {
    async fn handle_command(&mut self, command: ControllerCommand) {
        let result = match command {
            ControllerCommand::Connect => self.connect().await,
            ControllerCommand::Disconnect => {
                self.disconnect(true).await;
                Ok(())
            }
            ControllerCommand::Refresh => self.send_startup_queries().await,
            ControllerCommand::SetAnc(mode) => self.set_anc(mode).await,
            ControllerCommand::SetEq(mode) => self.set_eq(mode).await,
            ControllerCommand::SetCodec(mode) => self.set_codec(mode).await,
            ControllerCommand::SetGameMode(enabled) => self.set_game_mode(enabled).await,
            ControllerCommand::SetLowLatency(enabled) => self.set_low_latency(enabled).await,
        };
        if let Err(error) = result {
            self.emit_error(error.to_string());
        }
    }

    async fn connect(&mut self) -> Result<(), ControllerError> {
        if self.connection.is_some() {
            self.disconnect(false).await;
        }
        self.state.connection_state = ConnectionState::Connecting;
        self.state.last_error = None;
        self.emit_state();
        self.emit_log("正在搜索已配对的 NiceHCK 设备".to_owned());

        let candidates = timeout(DISCOVERY_TIMEOUT, transport::discover())
            .await
            .map_err(|_| ControllerError::DiscoveryTimeout)??;
        if candidates.is_empty() {
            self.connection_failed(ControllerError::NoCandidates.to_string());
            return Err(ControllerError::NoCandidates);
        }

        let mut failures = Vec::new();
        for candidate in candidates {
            self.emit_log(format!("尝试连接：{}", candidate.description()));
            match timeout(CONNECTION_TIMEOUT, transport::connect(candidate.clone())).await {
                Ok(Ok(connection)) => {
                    self.finish_connection(candidate, connection).await?;
                    return Ok(());
                }
                Ok(Err(error)) => {
                    let detail = format!("{}：{error}", candidate.description());
                    self.emit_log(format!("连接失败：{detail}"));
                    failures.push(detail);
                }
                Err(_) => {
                    let detail = format!("{}：连接超时", candidate.description());
                    self.emit_log(format!("连接失败：{detail}"));
                    failures.push(detail);
                }
            }
        }

        let detail = failures.join("；");
        self.connection_failed(detail.clone());
        Err(ControllerError::AllConnectionsFailed(detail))
    }

    async fn finish_connection(
        &mut self,
        candidate: Candidate,
        connection: Connection,
    ) -> Result<(), ControllerError> {
        self.state.connection_state = ConnectionState::Connected;
        self.state.device_name = candidate_device_name(&candidate).to_owned();
        self.state.port_name = Some(candidate_endpoint(&candidate));
        self.state.last_error = None;
        self.connection = Some(connection);
        self.emit_state();
        self.emit_log(format!("已连接：{}", candidate.description()));
        self.send_startup_queries().await
    }

    fn connection_failed(&mut self, message: String) {
        self.state.connection_state = ConnectionState::Error;
        self.state.last_error = Some(message);
        self.emit_state();
    }

    async fn disconnect(&mut self, announce: bool) {
        if let Some(connection) = self.connection.take()
            && let Err(error) = connection.disconnect().await
        {
            self.emit_log(format!("断开传输时出错：{error}"));
        }
        self.parser = PacketStreamParser::default();
        self.state = DeviceState::default();
        self.emit_state();
        if announce {
            self.emit_log("已断开连接".to_owned());
        }
    }

    async fn send_startup_queries(&mut self) -> Result<(), ControllerError> {
        self.ensure_connected()?;
        for packet in paced_startup_queries() {
            self.send_packet(packet).await?;
            sleep(COMMAND_PACING).await;
        }
        self.emit_log("已发送状态查询".to_owned());
        Ok(())
    }

    async fn set_anc(&mut self, mode: AncMode) -> Result<(), ControllerError> {
        self.ensure_connected()?;
        self.send_packet(set_anc(mode)).await?;
        sleep(COMMAND_PACING).await;
        self.send_packet(query_anc()).await?;
        self.emit_log(format!("已发送 ANC 切换：{}", mode.label()));
        Ok(())
    }

    async fn set_eq(&mut self, mode: EqMode) -> Result<(), ControllerError> {
        self.ensure_connected()?;
        if !self.state.firmware.supports_extended_eq()
            && matches!(mode, EqMode::Fine | EqMode::Vocal)
        {
            return Err(ControllerError::UnsupportedEq);
        }
        self.send_packet(set_eq(mode)).await?;
        sleep(COMMAND_PACING).await;
        self.send_packet(query_eq()).await?;
        self.emit_log(format!("已发送 EQ 切换：{}", mode.label()));
        Ok(())
    }

    async fn set_codec(&mut self, mode: CodecMode) -> Result<(), ControllerError> {
        self.ensure_connected()?;
        let packet = if self.state.firmware.supports_modern_codec_switch() {
            set_codec(mode)
        } else {
            if mode == CodecMode::Sbc {
                return Err(ControllerError::UnsupportedCodec);
            }
            set_legacy_codec_lhdc(mode == CodecMode::Lhdc)
        };
        self.send_packet(packet).await?;
        self.state.selected_codec = Some(mode);
        self.emit_state();
        self.emit_log(format!("已发送编码切换：{}", mode.label()));
        Ok(())
    }

    async fn set_game_mode(&mut self, enabled: bool) -> Result<(), ControllerError> {
        self.ensure_connected()?;
        self.send_packet(set_game_mode(enabled)).await?;
        sleep(COMMAND_PACING).await;
        self.send_packet(query_game_mode()).await?;
        self.emit_log(format!(
            "已发送游戏模式切换：{}",
            if enabled { "开" } else { "关" }
        ));
        Ok(())
    }

    async fn set_low_latency(&mut self, enabled: bool) -> Result<(), ControllerError> {
        self.ensure_connected()?;
        self.send_packet(set_low_latency(enabled)).await?;
        sleep(COMMAND_PACING).await;
        self.send_packet(query_low_latency()).await?;
        self.emit_log(format!(
            "已发送低延迟切换：{}",
            if enabled { "开" } else { "关" }
        ));
        Ok(())
    }

    fn ensure_connected(&self) -> Result<(), ControllerError> {
        if self.connection.is_some() {
            Ok(())
        } else {
            Err(ControllerError::NotConnected)
        }
    }

    async fn send_packet(&self, packet: Vec<u8>) -> Result<(), ControllerError> {
        let connection = self
            .connection
            .as_ref()
            .ok_or(ControllerError::NotConnected)?;
        connection.send(packet.clone()).await?;
        self.emit_log(format!("发送：{}", format_bytes(&packet)));
        Ok(())
    }

    async fn handle_transport_event(&mut self, event: Option<ConnectionEvent>) {
        match event {
            Some(ConnectionEvent::Data(data)) => {
                self.emit_log(format!("接收：{}", format_bytes(&data)));
                for message in self.parser.feed(&data) {
                    self.apply_update(parse_state_update(&message));
                }
            }
            Some(ConnectionEvent::Error(message)) => {
                self.emit_error(format!("蓝牙通信错误：{message}"));
            }
            Some(ConnectionEvent::Disconnected) | None => {
                self.connection = None;
                self.state.connection_state = ConnectionState::Error;
                self.state.last_error = Some("设备连接已断开".to_owned());
                self.emit_state();
                self.emit_error("设备连接已断开".to_owned());
            }
        }
    }

    fn apply_update(&mut self, update: ParsedStateUpdate) {
        let mut changed = false;
        if let Some(firmware) = update.firmware {
            self.state.firmware = firmware;
            changed = true;
        }
        if let Some(value) = update.left_battery {
            self.state.left_battery = Some(value);
            changed = true;
        }
        if let Some(value) = update.right_battery {
            self.state.right_battery = Some(value);
            changed = true;
        }
        if let Some(value) = update.case_battery {
            self.state.case_battery = value;
            changed = true;
        }
        if let Some(value) = update.anc_mode {
            self.state.anc_mode = value;
            changed = true;
        }
        if let Some(value) = update.eq_mode {
            self.state.eq_mode = value;
            changed = true;
        }
        if let Some(value) = update.game_mode_enabled {
            self.state.game_mode_enabled = Some(value);
            changed = true;
        }
        if let Some(value) = update.low_latency_enabled {
            self.state.low_latency_enabled = Some(value);
            changed = true;
        }
        if changed {
            self.emit_state();
        }
    }

    fn emit_state(&self) {
        let _ = self.events.send(ControllerEvent::State(self.state.clone()));
    }

    fn emit_log(&self, message: String) {
        info!("{message}");
        let _ = self.events.send(ControllerEvent::Log(message));
    }

    fn emit_error(&self, message: String) {
        error!("{message}");
        let _ = self.events.send(ControllerEvent::Error(message));
    }
}

fn candidate_device_name(candidate: &Candidate) -> &str {
    match candidate {
        Candidate::Rfcomm(candidate) => &candidate.device_name,
        Candidate::Serial(candidate) => &candidate.device_name,
    }
}

fn candidate_endpoint(candidate: &Candidate) -> String {
    match candidate {
        Candidate::Rfcomm(candidate) => format!("RFCOMM:{}", candidate.address),
        Candidate::Serial(candidate) => candidate.port_name.clone(),
    }
}

fn format_bytes(bytes: &[u8]) -> String {
    bytes
        .iter()
        .map(|byte| format!("{byte:02X}"))
        .collect::<Vec<_>>()
        .join(" ")
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::models::FirmwareVersion;

    #[test]
    fn formats_packets_for_diagnostics() {
        assert_eq!(format_bytes(&[0x4e, 0, 0xff]), "4E 00 FF");
    }

    #[test]
    fn applies_explicitly_unknown_case_battery() {
        let (events, _) = mpsc::unbounded_channel();
        let mut actor = ControllerActor {
            state: DeviceState {
                case_battery: Some(50),
                ..DeviceState::default()
            },
            parser: PacketStreamParser::default(),
            connection: None,
            events,
        };
        actor.apply_update(ParsedStateUpdate {
            case_battery: Some(None),
            ..ParsedStateUpdate::default()
        });
        assert_eq!(actor.state.case_battery, None);
    }

    #[test]
    fn firmware_capability_boundary_matches_existing_app() {
        assert!(!FirmwareVersion::new(4, 7).supports_extended_eq());
        assert!(FirmwareVersion::new(4, 8).supports_extended_eq());
    }
}
